tests/cpp/test_multidevice_gpu_comms.cpp

```diff
@@ -0,0 +1,151 @@
+#include <host_ir/container.h>
+#include <host_ir/executor.h>
+#include <ir/all_nodes.h>
+#include <ops/all_ops.h>
+#include <tests/cpp/multidevice.h>
+#include <tests/cpp/multidevice_kernels.h>
+
+namespace nvfuser {
+
+#define CUDA_CALL(call) ASSERT_EQ((call), cudaSuccess)
+
+class GpuCommTest : public MultiDeviceTest {};
+
+TEST_F(GpuCommTest, IpcMemHandle) {
+  // Allocate GPU memory
+  constexpr size_t size = sizeof(int64_t);
+  const int64_t num_devices = communicator_->size();
+  const int64_t rank = communicator_->deviceId();
+  void* d_ptr;
+  CUDA_CALL(cudaMalloc(&d_ptr, size));
+
+  const int64_t value = rank;
+  CUDA_CALL(cudaMemcpy(d_ptr, &value, size, cudaMemcpyHostToDevice));
+
+  cudaIpcMemHandle_t ipc_handle;
+  CUDA_CALL(cudaIpcGetMemHandle(&ipc_handle, d_ptr));
+
+  auto store = communicator_->getTcpStore();
+  store->set("ipc_handle_" + std::to_string(rank), toBytes(ipc_handle));
+  communicator_->barrier();
+  auto peer_ipc_handle = fromBytes<cudaIpcMemHandle_t>(store->get("ipc_handle_" + std::to_string((rank + 1) % num_devices)));
+
+  void* peer_d_ptr;
+  CUDA_CALL(cudaIpcOpenMemHandle(&peer_d_ptr, peer_ipc_handle, cudaIpcMemLazyEnablePeerAccess));
+
+  int64_t peer_value;
+  CUDA_CALL(cudaMemcpy(&peer_value, peer_d_ptr, size, cudaMemcpyDeviceToHost));
+
+  EXPECT_EQ((value + 1) % num_devices, peer_value);
+
+  // Clean up
+  CUDA_CALL(cudaIpcCloseMemHandle(peer_d_ptr));
+  CUDA_CALL(cudaFree(d_ptr));
+
+}
+
+TEST_F(GpuCommTest, IpcMemHandlePtrArithmeticAtReceiver) {
+  // TLDR; We can do pointer arithmetic on the receiver side.
+
+  // Allocate GPU memory
+  constexpr size_t size = 2 * sizeof(int64_t);
+  const int64_t num_devices = communicator_->size();
+  const int64_t rank = communicator_->deviceId();
+  const int64_t peer_rank = (rank + 1) % num_devices;
+  void* d_ptr;
+  CUDA_CALL(cudaMalloc(&d_ptr, size));
+
+  std::vector<int64_t> values;
+  values.push_back(2 * rank);
+  values.push_back(2 * rank + 1);
+  CUDA_CALL(cudaMemcpy(d_ptr, values.data(), size, cudaMemcpyHostToDevice));
+
+  cudaIpcMemHandle_t ipc_handle;
+  CUDA_CALL(cudaIpcGetMemHandle(&ipc_handle, d_ptr));
+
+  auto store = communicator_->getTcpStore();
+  store->set("ipc_handle_" + std::to_string(rank), toBytes(ipc_handle));
+  communicator_->barrier();
+  auto peer_ipc_handle = fromBytes<cudaIpcMemHandle_t>(store->get("ipc_handle_" + std::to_string(peer_rank)));
+
+  int64_t* peer_d_ptr;
+  CUDA_CALL(cudaIpcOpenMemHandle((void**)&peer_d_ptr, peer_ipc_handle, cudaIpcMemLazyEnablePeerAccess));
+
+  int64_t peer_value;
+  CUDA_CALL(cudaMemcpy(&peer_value, peer_d_ptr + 1, size / 2, cudaMemcpyDeviceToHost));
+
+  EXPECT_EQ(2 * peer_rank + 1, peer_value);
+
+  // Clean up
+  CUDA_CALL(cudaIpcCloseMemHandle(peer_d_ptr));
+  CUDA_CALL(cudaFree(d_ptr));
+}
+
+TEST_F(GpuCommTest, IpcMemHandlePtrArithmeticAtSender) {
+  // TLDR; We CANNOT do pointer arithmetic on the sender side! The IPC handle points to the beginning of the allocated buffer.
```

Comment:
@syed-ahmed I know it's a long shot, but does this ring a bell? I suspect it's related to some of the mempool work that you did. 

I'm afraid this can be a big problem when we use at::Tensor. AFAICT, an at::Tensor's buffer can be a slice of a larger cudaMalloc. If we can't send a slice, we can't send an at::Tensor reliably.

Reply:
This could be the answer, taken from [the doc](https://docs.nvidia.com/cuda/cuda-c-programming-guide/#interprocess-communication)

> Note that allocations made by cudaMalloc() may be sub-allocated from a larger block of memory for performance reasons. In such case, CUDA IPC APIs will share the entire underlying memory block which may cause other sub-allocations to be shared, which can potentially lead to information disclosure between processes. To prevent this behavior, it is recommended to only share allocations with a 2MiB aligned size.

Reply:
Yes, @samnordmann's reference is the correct one. If you want fine-grained control over the allocations, you can also use the cuMem style APIs and export/import handles for IPC: https://developer.nvidia.com/blog/introducing-low-level-gpu-virtual-memory-management/#operating_system_native_interprocess_communication.